### PR TITLE
Types overrides imports

### LINKED ISSUE #480: Improve typing customization
Once again, amazing work on your part @adelsz!

Here are a few ideas for improving typing with pgtyped. I'm ready to do it but I wanted to have everyone's aproval on the idea before going further.

# Complex mappers
Today it supports basic TS types:
```json
{
  "typesOverrides": {
    "date": "string",
    "json": "any"
  }
}
```

It would be nice to be able to specify imported types in the fashion of what [GraphQL codegen does](https://the-guild.dev/graphql/codegen/plugins/typescript/typescript-resolvers#use-your-model-types-mappers).

```json
{
  "typesOverrides": {
    "date": "dayjs#Dayjs"
    "myEnum": "./types/enums.ts#MyEnum"
  }
}
```

❓ What should be the root of the relative paths?
❓ What should be the specs of the format?

# Different mapping for parameters
It would be nice to type parameters and return values differently, even if they are of the same PG type.
For instance `Decimal` are returned as `string` by PG, so all select should be typed as such, but when inserting, updating, or using it in a where clause PG accept both a string or a number.

- The default typing should reflect that
- The config file should also allow to override those types as well

❓ What should the config file look like exactly?
❓ What is the behavior expected if you only override one of the two?

# Customize generated types
Today is generates `interface` and `enum`, it would be nice to choose between `type` `interface` and even `class` and between `enum` and string union.

 types as well

❓ What should the config file look like exactly?

### PR DESCRIPTION
**Pull request by nick-keller**
Fix #480 

Please review it carefully, I did not test it end to end.

### CODE REVIEW

File: docs-new/docs/typing.md [hunk 1]
```diff
@@ -0,0 +1,150 @@
...
+import { types } from 'pg'
+
+// DATE are now returned as string instead of Date objects
+types.setTypeParser(types.builtins.DATE, (val: string) => val);
+
+// DECIMAL and other precision types are now returned as number instead of string
+types.setTypeParser(types.builtins.NUMERIC, (val: string) => Number(val));
+```
+
+This is posrt of the [`pg` lib](https://github.com/brianc/node-pg-types) and has nothing to do with `pgtyped`.
```
Comment (adelsz) [inline docs-new/docs/typing.md:23]: typo?

File: docs-new/docs/typing.md [hunk 2]
```diff
@@ -0,0 +1,150 @@
...
+```
+
+You can also specify imported types from your project or from another npm package:
+```json
+{
+  "typesOverrides": { 
+    "timestamptz": "dayjs#Dayjs", // import { Dayjs } from 'dayjs';
+    "money": "my-package#Foo as MyType", // import { Foo as MyType } from 'my-package';
+    "char": "my-package as MyType", // import MyType from 'my-package';
+    "numeric": "./path/to/file#MyCustomType", // import { MyCustomType } from './path/to/file';
```
Comment (adelsz) [inline docs-new/docs/typing.md:45]: Lets add an example for an ESM style import (extension required):
```suggestion
    "numeric": "./path/to/file.js#MyCustomType", // import { MyCustomType } from './path/to/file.js';
```

File: packages/cli/src/declareImport.test.ts [hunk 3]
```diff
@@ -0,0 +1,80 @@
...
+        { name: 'Alias2', from: 'package', aliasOf: 'Baz' },
+      ],
+      './',
+    ),
+  ).toBe(
+    "import Alias, { Foo as Alias1, Bar, Baz as Alias2 } from 'package';\n",
+  );
+});
+
+describe('relative imports', () => {
```
Comment (adelsz) [inline packages/cli/src/declareImport.test.ts:53]: Lets add some tests for ESM style imports

File: packages/cli/src/types.ts [hunk 4]
```diff
@@ -129,7 +220,7 @@ export class TypeAllocator {
   }
 
   /** Lookup a database-provided type name in the allocator's map */
-  use(typeNameOrType: MappableType): string {
+  use(typeNameOrType: MappableType, scope: 'parameter' | 'return'): string {
```
Comment (adelsz) [inline packages/cli/src/types.ts:223]: Lets introduce an enum for scope instead of string union.

**Review by adelsz — APPROVED** [review]
Thanks for this PR @nick-keller!
I added some comments mainly regarding `scope` not being an enum and a few good to have ESM test cases.
Will you be open to adding some e2e test cases to `packages/example`?

**Comment by nick-keller** [comment]
Hey @adelsz,
I took your comments into consideration but I could not make the example work.

Running `test` gives me:
```
Step 2/6 : FROM $NODE_IMAGE
ERROR: Service 'build' failed to build : invalid reference format
```

And running `check` gives me:
```
error TS6059: File '/Users/nick/projects/pgtyped/packages/example/jest-cjs.config.ts' is not under 'rootDir' '/Users/nick/projects/pgtyped/packages/example/src'. 'rootDir' is expected to contain all source files.
  The file is in the program because:
    Matched by default include pattern '**/*'

error TS6059: File '/Users/nick/projects/pgtyped/packages/example/jest.config.ts' is not under 'rootDir' '/Users/nick/projects/pgtyped/packages/example/src'. 'rootDir' is expected to contain all source files.
  The file is in the program because:
    Matched by default include pattern '**/*'


Found 2 errors.
```

If you can help me make it work so that I can add examples, or add them yourself if you feel like it that;s be great!

**Comment by adelsz** [comment]
Thanks @nick-keller. The changes look good.
To run the example packages you need to specify the node version as follows `NODE_VERSION=16 npm run test`.
We run it as part of the CI https://github.com/adelsz/pgtyped/blob/master/.github/workflows/main.yml

**Comment by nick-keller** [comment]
Sorry @adelsz but I cannot get it to work.
Running `check` still gives me the two errors I listed above. And running `test` kinda works until `src/index.test.ts` timesout. This is definitly not my cup of tea, and I don't have the skills and the patience to solve those kind of issues.

Furthermore:
- The tests are not really checking that the generated types are correct. For instance `expect(typeof dateAsString).toBe('string');` only checks that `pg.types.setTypeParser(...)` works, it is not checking that **pgtyped** actually typed the column as string.
- In the package.json, dependencies are `^2.0.0`, meaning that my work will not be reflected until you actually publish the package anyway.

I'll have to let you take it from here I'm afraid :/

**Comment by adelsz** [comment]
No worries, I can handle that on my side. Thanks for working on this major feature, it will be a great addition to pgTyped.

As for the issues you encountered, `packages/example` is one of the confusing parts of the codebase. I will have to write more detailed instructions in a `CONTRIBUTING.md` doc.

In case it might be of any future use I will comment on the issues you brought up:

> The tests are not really checking that the generated types are correct. For instance expect(typeof dateAsString).toBe('string'); only checks that pg.types.setTypeParser(...) works, it is not checking that pgtyped actually typed the column as string.

The idea of the `example` package is to serve both as an e2e test (compile-time and runtime) and example usage of pgTyped.

The `npm run test` command does two things:
1. Runs pgTyped, generating code for all queries. After running the type generation, a `git diff` will show if there are any unexpected problems with the generated queries and types (e2e compile-time test).
2. Runs the generated queries as part of a Jest test suite. This ensures that the queries work at runtime.

> In the package.json, dependencies are ^2.0.0, meaning that my work will not be reflected until you actually publish the package anyway.

pgTyped is a monorepo using npm workspaces, so running `npm install` actually crosslinks all of the `pgTyped` packages together locally (the exact versions in package jsons don't matter). Running `npm run build` builds all packages and makes them available to each other as if they were published.

As I said, all of this belongs in `CONTRIBUTING.md` doc, which has been long overdue.

**Pull request merged**